### ISSUE TITLE
Client sometimes crashes on world transition

### ISSUE THREAD

**Issue by minexew** [Pokechu22/WorldDownloader#73]
This time it happened upon entering nether. Not sure if I encountered it the other way as well.
MC 1.12.2, Forge, litemod loader, WDL

```
java.lang.RuntimeException: No OpenGL context found in the current thread.
	at org.lwjgl.opengl.GLContext.getCapabilities(GLContext.java:124)
	at org.lwjgl.opengl.GL30.glBindFramebuffer(GL30.java:1122)
	at net.minecraft.client.renderer.OpenGlHelper.func_153171_g(OpenGlHelper.java:665)
	at net.minecraft.client.shader.Framebuffer.func_147609_e(Framebuffer.java:215)
	at net.minecraft.client.shader.Framebuffer.func_147608_a(Framebuffer.java:68)
	at net.minecraft.client.shader.ShaderGroup.func_148021_a(ShaderGroup.java:331)
	at net.minecraft.client.renderer.EntityRenderer.func_175066_a(EntityRenderer.java:301)
	at net.minecraft.client.Minecraft.func_175607_a(Minecraft.java:3138)
	at wdl.gui.GuiTurningCameraBase.func_146281_b(GuiTurningCameraBase.java:175)
	at net.minecraft.client.Minecraft.func_147108_a(Minecraft.java:987)
	at wdl.WDL.startDownload(WDL.java:413)
	at wdl.WDL.loadWorld(WDL.java:547)
	at wdl.WDL.onSaveComplete(WDL.java:568)
	at wdl.WDL.lambda$startSaveThread$0(WDL.java:475)
	at wdl.WDL$$Lambda$214/1535894771.run(Unknown Source)
	at java.lang.Thread.run(Thread.java:745)
```

Detailed error dump: https://gist.github.com/minexew/bb8582b6412fd0aeef70f4719fb0d3f1

**Comment by Pokechu22** [comment]
Hm, probably related to #72.  Though, that actual stacktrace is weird, it's related to shaders.  Here's the same thing, remapped:

```
java.lang.RuntimeException: No OpenGL context found in the current thread.
	at org.lwjgl.opengl.GLContext.getCapabilities(GLContext.java:124)
	at org.lwjgl.opengl.GL30.glBindFramebuffer(GL30.java:1122)
	at net.minecraft.client.renderer.OpenGlHelper.glBindFramebuffer(OpenGlHelper.java:665)
	at net.minecraft.client.shader.Framebuffer.unbindFramebuffer(Framebuffer.java:215)
	at net.minecraft.client.shader.Framebuffer.deleteFramebuffer(Framebuffer.java:68)
	at net.minecraft.client.shader.ShaderGroup.deleteShaderGroup(ShaderGroup.java:331)
	at net.minecraft.client.renderer.EntityRenderer.loadEntityShader(EntityRenderer.java:301)
	at net.minecraft.client.Minecraft.setRenderViewEntity(Minecraft.java:3138)
	at wdl.gui.GuiTurningCameraBase.onGuiClosed(GuiTurningCameraBase.java:175)
	at net.minecraft.client.Minecraft.displayGuiScreen(Minecraft.java:987)
	at wdl.WDL.startDownload(WDL.java:413)
	at wdl.WDL.loadWorld(WDL.java:547)
	at wdl.WDL.onSaveComplete(WDL.java:568)
	at wdl.WDL.lambda$startSaveThread$0(WDL.java:475)
	at wdl.WDL$$Lambda$214/1535894771.run(Unknown Source)
	at java.lang.Thread.run(Thread.java:745)
```

You do have optifine installed, so I'm tempted to blame that, but I'm not completely sure that it's actually at fault.  WDL does change the render view entity during that screen, to have the panoramic background, and when you change that the game has the option of loading a different shader (in spectator mode, if you spectate a spider for instance).  But I wouldn't expect it to behave like that.

**Comment by Pokechu22** [comment]
This is what `EntityRenderer.loadEntityShader` looks like:

```java
	/**
	 * What shader to use when spectating this entity
	 */
	public void loadEntityShader(@Nullable Entity entityIn) {
		if (OpenGlHelper.shadersSupported) {
			if (this.shaderGroup != null) {
				this.shaderGroup.deleteShaderGroup();
			}

			this.shaderGroup = null;

			if (entityIn instanceof EntityCreeper) {
				this.loadShader(new ResourceLocation("shaders/post/creeper.json"));
			} else if (entityIn instanceof EntitySpider) {
				this.loadShader(new ResourceLocation("shaders/post/spider.json"));
			} else if (entityIn instanceof EntityEnderman) {
				this.loadShader(new ResourceLocation("shaders/post/invert.json"));
			}
		}
	}
```

Since `deleteShaderGroup` was called, `shaderGroup` must not be null.  However, it should have already been null when the GUI was opened (since I change the render view entity then), unless optifine does some extra logic for shaders (which would make sense).  Did you have a shader active with optifine?

In any case, I do see the issue on my end; I'm changing the render view entity unsafely from the wrong thread.  Should be easy to fix, but I won't be able to test it.

**Comment by minexew** [comment]
Nope, I wasn't using any shaders at the time.

**Comment by minexew** [comment]
I'll disable OptiFine for now to see what happens.

**Comment by Pokechu22** [comment]
OK, I've pushed a change that I _think_ should fix this.  As with #72 you won't be able to use it until I actually make a release (unless you compile the mod yourself), though; I'm still in the middle of another refactor which means I can't make a stable release just yet.

Investigating multiworld and all of the other world management stuff is something that I've been meaning to do, since I'm sure there's a ton of other subtle bugs around it that just haven't been found yet.  But that won't be for until the release after the next; right now I'm busy with block entity handlers (which are a more pressing concern for preparing for 1.13).

**Comment by minexew** [comment]
Thanks. Indeed after disabling OptiFine the error doesn't seem to occur, but I haven't played long enough to be sure about it.

**Comment by Pokechu22** [comment]
I've released [version 4.0.2.0](https://github.com/Pokechu22/WorldDownloader/releases/tag/v4.0.2.0) which should fix this issue.  Please let me know if it's still happening after updating, and I'll reopen this issue.
